modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SSource.java

```diff
@@ -312,7 +322,14 @@ void write(MethodWriter writer, Globals globals) {
         }
 
         if (!methodEscape) {
-            writer.visitInsn(Opcodes.ACONST_NULL);
+            if (scriptInterface.getExecuteMethodReturnType().equals(Definition.DEF_TYPE)) {
+                // TODO this should fail too but it can't to keep compatibility with Elasticsearch scripts
+                writer.visitInsn(Opcodes.ACONST_NULL);
+            } else if (false == scriptInterface.getExecuteMethodReturnType().equals(Definition.VOID_TYPE)) {
+                assert false : "this should have failed during the analyze phase";
+                throw new IllegalArgumentException("Expected all paths to [return] but not all did or end in an expression to return but "
+                        + "some path ends is missing a return and ends in a statement.");
+            }
```

Comment:
^^ Seems like we agree based on your comments made while I was making this one :)